app/services/steam_import_service.rb

```diff
@@ -0,0 +1,108 @@
+# typed: true
+require 'set'
+
+class SteamImportService
+  extend T::Sig
+
+  sig { params(user: User).void }
+  def initialize(user)
+    @user = user
+  end
+
+  class Error < StandardError; end
+  class NoGamesError < Error; end
+
+  sig { returns(Result) }
+  def call
+    raise Error, 'No Steam account.' if steam_account.nil?
+
+    json = JSON.parse(T.must(T.must(URI.open(steam_api_url)).read))
+
+    games = json.dig('response', 'games')
+
+    raise NoGamesError if games.nil?
+
+    games.reject! { |game| game['img_logo_url'].blank? }
+
+    steam_ids = T.cast(
+      games.map { |g| g['appid'].to_i }.uniq,
+      T::Array[Integer]
+    )
+    stored_ids = T.cast(SteamAppId.where(app_id: steam_ids).pluck(:app_id, :game_id), T::Array[[Integer, Integer]])
+    stored_id_hash = stored_ids.each_with_object({}) do |(app_id, game_id), hash|
+      hash[app_id] = game_id
+    end
+    missing_ids = T.cast(
+      steam_ids.to_set - stored_ids.map(&:first).to_set,
+      T::Set[Integer]
+    )
+
+    create_time = Time.current
+
+    attrs =
+      games.each
+           .lazy
+           .reject { |game| missing_ids.include?(game['appid']) }
+           .map do |game_info|
+        hours_played = (game_info['playtime_forever'].to_f / 60).round(1)
+        {
+          hours_played: hours_played,
+          game_id: stored_id_hash[game_info['appid']],
+          user_id: user.id,
+          created_at: create_time,
+          updated_at: create_time
+        }
+      end
+    attrs = attrs.to_a
+    added_purchases =
+      if attrs.any?
+        inserted = GamePurchase.insert_all(attrs, unique_by: [:game_id, :user_id])
```

Comment:
this doesn't perform validations and that scares me a little

But more importantly, this doesn't perform callbacks >:(

Reply:
You already disable the activity stuff by disabling bullet.

It's this or the code is slow AF okay

Reply:
Disabling bullet doesn't do anything to validations/callbacks afaik?

Reply:
Oh right. I can make this `.find_or_create_by`, which is still *extremely* slow compared to a big upsert, but it'll still be faster than the current scheme.